Jakarta Annotations 2.0

m0mus opened the pull request:
Signed-off-by: Dmitry Kornilov <dmitry.kornilov@oracle.com>

## Specification PR template
When creating a specification project release review, create PRs with the content defined as follows.

Include the following in the PR:
- [x] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [x] Specification PDF in the form of wombat-spec-x.y.pdf
- [x] Specification HTML in the form of wombat-spec-x.y.html
- [x] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [x] For a Progress Review, that sufficient progress has been made on a Compatible Implementation and TCK, to ensure that the spec is implementable and testable.
- [x] For a Release Review, a summary that a Compatible Implementation is complete, passes the TCK, and that the TCK includes sufficient coverage of the specification.
The TCK users guide MUST include the instructions to run the compatible implementations used to validate the release.
Instructions MAY be by reference.
- [x] The URL of the OSSRH staging repository for the api, javadoc:
      https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/annotation/jakarta.annotation-api/2.0.0/
- [x] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted/jakarta-annotations-tck-2.0.0.zip
- [x] The URL of the compatibility certification request issue:
      https://github.com/eclipse-ee4j/common-annotations-api/issues/81
- [x] Specification JavaDoc in the wombat/x.y/apidocs directory. 
If desired, an optional second PR can be created to contain just the JavaDoc in the `apidocs` directory.

kwsutter commented:
According to the [Initial Reviewer list](https://docs.google.com/document/d/1KaSQBfiyJc2whfQdeuQ1Y7mXnnUnnCuiE-nAqUq5Foo/edit), @edbratt and @m0mus are supposed to help review the Annotation PRs.  But, since Dmitry is creating them, I will take a first shot at the review.  Will need some help with follow-up though since I am out all of next week (July 16-24).

kwsutter commented:
# Spec Review Checklist

1. Spec PR
  - [ ] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [ ] Directory of form {spec}/x.y
  - [ ] PDF of form {spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [ ] HTML of form {spec}-spec-x.y.html ("-spec" preferred but not required)
  - [ ] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [ ] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/{spec}/jakarta-{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [ ] (Optional) Second PR for just apidocs

2. _index.md
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT
  - [ ] Correct Logo

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request
 
10. Update Jakarta EE API jar
  - [ ] Update the Jakarta EE API jar by submitting a PR to the jakartaee-api project that updates the version number of your API jar file.

kwsutter commented on annotations/2.0/_index.md:
```diff
@@ -1,23 +1,23 @@
 ---
-title: "Jakarta Annotations (milestone)"
-date: 2020-05-31
+title: "Jakarta Annotations"
```
Needs a version --- Jakarta Annotations 2.0

m0mus replied:
Fixed

kwsutter commented on annotations/2.0/_index.md:
```diff
@@ -1,23 +1,23 @@
+* [Jakarta Annotations 2.0 Specification Document](./annotations-spec-2.0.html) (HTML)
 * [Jakarta Annotations 2.0 Javadoc](./apidocs)
-* [Jakarta Annotations 2.0 TCK]()([sig](),[sha](),[pub]())
+* [Jakarta Annotations 1.3 TCK](https://download.eclipse.org/jakartaee/annotations/2.0/jakarta-annotations-tck-2.0.0.zip) ([sig](https://download.eclipse.org/jakartaee/annotations/2.0/jakarta-annotations-tck-2.0.0.zip.sig),[sha](https://download.eclipse.org/jakartaee/annotations/2.0/jakarta-annotations-tck-2.0.0.zip.sha256),[pub](https://raw.githubusercontent.com/jakartaee/specification-committee ...
```
Jakarta Annotations 2.0 TCK (not 1.3).

m0mus replied:
Fixed.

kwsutter commented on annotations/2.0/_index.md:
```diff
@@ -1,23 +1,23 @@
 # Compatible Implementations
 
-* [Eclipse GlassFish]()
+* [Eclipse GlassFish](https://eclipse-ee4j.github.io/glassfish/download)
```
Does Annotation have it's own standalone CI?  If not, how do we merge this before the final Glassfish is available?

m0mus replied:
It's tricky. Jakarta Annotations doesn't have a reference (compatible) implementation because it's just a set of annotations. Also, there are no standalone TCKs, just a signature test. I put there GlassFish because it was listed as a compatible implementation in version 1.3. Putting GlassFish is fine because it implements an annotation processor for Jakarta Annotations API, but I am not sure that it's a compatible implementation in terms how we understand it.
I am suggesting removing this section and possibly add it later when GF6 is released. Thoughts?

kwsutter replied:
Based on what we did with 1.3, the following url was used for the CI:  https://eclipse-ee4j.github.io/common-annotations-api/  Given the unique case for Annotations, I'd be fine with this link for the 2.0 CI as well.

m0mus replied:
Fixed

kwsutter commented on annotations/2.0/apidocs/META-INF/MANIFEST.MF:
```diff
@@ -1,5 +1,2 @@
 Manifest-Version: 1.0
```
We should not be including the META-INF directory in the javadocs.  This comment would apply to all three files specified in this PR.

m0mus replied:
Fixed.

kwsutter commented:
Thanks, @m0mus, for the updates.  I'm still looking for the TCK zip file and the Compatibility Certification request.

m0mus commented:
> Thanks, @m0mus, for the updates. I'm still looking for the TCK zip file and the Compatibility Certification request.
Working on it.

edbratt commented:
I created a place-holder for the TCK Results page, [here](https://eclipse-ee4j.github.io/glassfish/certifications/jakarta-annotations/2.0/TCK-Results). You are welcome to leave this with the GlassFish implementation, or use it to create another results page. It will need to be refreshed when the TCK is updated to contain the revised documentation (merged earlier today). [to confirm the results are still good and update the SHA]. Then, you can submit the Compatibility Certification Request issue in the [eclipse-ee4j/common-annotations-api/issues](https://github.com/eclipse-ee4j/common-annotations-api/issues/)

alwin-joseph commented:
The TCK has been updated and released to https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted/jakarta-annotations-tck-2.0.0.zip . Can you please use the same for the certification.

kwsutter commented:
> The TCK has been updated and released to https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted/jakarta-annotations-tck-2.0.0.zip . Can you please use the same for this PR.

Thanks, @alwin-joseph!  Will use that for a review.

kwsutter commented:
The TCK looks good.  Now, we just need the Certification Request with the TCK Results and we should be ready to go.  Thanks!

kwsutter commented:
@m0mus, will there be a Certification Request soon?  This one seems very close to completing...  Thanks!

kwsutter commented:
@m0mus bump...

m0mus commented:
@kwsutter Finished. Sorry for delay.

kwsutter commented:
@m0mus The SHA values don't match up between the CR and the TCK Results.  I'm assuming the more accurate TCK is in the "promoted" directory, but I need you to verify before I start a vote.  Thanks!

kwsutter reviewed (commented):
Comment added...

m0mus commented:
SHA is fixed.

kwsutter reviewed (approved):
LGTM; Thanks, @m0mus!

kwsutter commented:
Not sure why some of @m0mus's commits are being flagged as not being signed.  I'm going to go ahead with the ballot and maybe we can get this cleaned up before we merge.

kwsutter commented:
@m0mus Based on the discussion on today's Spec Committee call, would you care to update your reference to the Glassfish Compatible Implementation?  Maybe find another, more permanent location for the nightly driver you used?  Thanks!

m0mus commented:
@kwsutter One commit was not signed off. It caused problems with Eclipse ECA checker. I squashed everything to one commit and changed a compatible implementation to GF 6.0.0-M1. TCKs are passed on this version too and it's released to Maven Central.

kwsutter commented:
> @kwsutter One commit was not signed off. It caused problems with Eclipse ECA checker. I squashed everything to one commit and changed a compatible implementation to GF 6.0.0-M1. TCKs are passed on this version too and it's released to Maven Central.

I just noticed that the TCK Results doesn't have a Compatible Implementation reference.  I don't know if it was always missing, or if something accidentally got deleted.  But, the TCK Results is supposed to have the same information as the CR.  Can you update the TCK Results with a reference to the Glassfish M1 binary?  Thanks!

lukasj commented:
this is currently blocked by: https://issues.redhat.com/browse/WELD-2641

manovotn commented:
@lukasj I have created a PR for that Weld issue here - https://github.com/weld/core/pull/2154
Can you confirm this fixes the issue for you?
Also, I suppose you are going to need Weld release with this fix in order to proceed, correct?

lukasj commented:
@manovotn Thanks for the quick fix, Matěji. While I was testing it (my 4.0.0 SNAPSHOT from weld/core corresponds to `39eb51b`), I've noticed https://issues.redhat.com/browse/WELD-2642 which was hidden...

manovotn commented:
Darn it, I already staged the release. I will check it tomorrow and look into fixing that as well.

manovotn commented:
@lukasj I have released Weld 4.0.0.Beta2 with those two fixes, it should land in Central any time now.

arjantijms commented:
@manovotn 
It was already discussed elsewhere, but for completion I'll mention it here too: Weld 4.0.0.Beta2 unfortunately doesn't work. It includes among others Servlet classes, which causes the bundle to fail on at least GlassFish.

manovotn commented:
@arjantijms noted, fix for reported issue (WELD-2644) was merged and Weld 4.0.0.Beta3 is now released.

arjantijms commented:
@manovotn 
> noted, fix for reported issue (WELD-2644) was merged and Weld 4.0.0.Beta3 is now released.

Thanks, although unfortunately Beta3 doesn't work either :(

Still tons of exceptions when running the GF `cdi_all` test like this:

```
     [exec] Caused by: javax.naming.NamingException: Lookup failed for 'java:global/simple-ejb-cdiApp/simple-ejb-cdi-ejb/SfulEJB!com.sun.s1asdev.cdi.ejb.hello.session.Sful' in SerialContext[myEnv={java.naming.factory.initial=com.sun.enterprise.naming.impl.SerialInitContextFactory, java.naming.factory.url.pkgs=com.sun.enterprise.naming, java.naming.factory.state=com.sun.corba.ee.impl.presentation.rmi.JNDIStateFactoryImpl} [Root exception is javax.naming.NamingException: ejb ref resolution error for remote business interfacecom.sun.s1asdev.cdi.ejb.hello.session.Sful [Root exception is java.lang.reflect.UndeclaredThrowableException]]
     [exec] 	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:467)
     [exec] 	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:414)
     [exec] 	at javax.naming.InitialContext.lookup(InitialContext.java:417)
     [exec] 	at com.sun.ejb.EjbNamingReferenceManagerImpl.resolveEjbReference(EjbNamingReferenceManagerImpl.java:159)
     [exec] 	... 23 more
     [exec] Caused by: javax.naming.NamingException: ejb ref resolution error for remote business interfacecom.sun.s1asdev.cdi.ejb.hello.session.Sful [Root exception is java.lang.reflect.UndeclaredThrowableException]
     [exec] 	at com.sun.ejb.EJBUtils.lookupRemote30BusinessObject(EJBUtils.java:409)
     [exec] 	at com.sun.ejb.containers.RemoteBusinessObjectFactory.getObjectInstance(RemoteBusinessObjectFactory.java:51)
     [exec] 	at javax.naming.spi.NamingManager.getObjectInstance(NamingManager.java:321)
     [exec] 	at com.sun.enterprise.naming.impl.SerialContext.getObjectInstance(SerialContext.java:503)
     [exec] 	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:463)
     [exec] 	... 26 more
     [exec] Caused by: java.lang.reflect.UndeclaredThrowableException
     [exec] 	at com.sun.proxy.$Proxy285.create(Unknown Source)
     [exec] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
     [exec] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
     [exec] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
     [exec] 	at java.lang.reflect.Method.invoke(Method.java:498)
     [exec] 	at com.sun.corba.ee.impl.presentation.rmi.ReflectiveTie.dispatchToMethod(ReflectiveTie.java:113)
     [exec] 	at com.sun.corba.ee.impl.presentation.rmi.ReflectiveTie._invoke(ReflectiveTie.java:143)
     [exec] 	at com.sun.corba.ee.impl.protocol.ServerRequestDispatcherImpl.dispatchToServant(ServerRequestDispatcherImpl.java:492)
     [exec] 	at com.sun.corba.ee.impl.protocol.ServerRequestDispatcherImpl.dispatch(ServerRequestDispatcherImpl.java:163)
     [exec] 	at com.sun.corba.ee.impl.protocol.MessageMediatorImpl.handleRequestRequest(MessageMediatorImpl.java:1520)
     [exec] 	at com.sun.corba.ee.impl.protocol.MessageMediatorImpl.handleRequest(MessageMediatorImpl.java:1396)
     [exec] 	at com.sun.corba.ee.impl.protocol.MessageMediatorImpl.handleInput(MessageMediatorImpl.java:901)
     [exec] 	at com.sun.corba.ee.impl.protocol.giopmsgheaders.RequestMessage_1_2.callback(RequestMessage_1_2.java:183)
     [exec] 	at com.sun.corba.ee.impl.protocol.MessageMediatorImpl.handleRequest(MessageMediatorImpl.java:665)
     [exec] 	at com.sun.corba.ee.impl.protocol.MessageMediatorImpl.dispatch(MessageMediatorImpl.java:467)
     [exec] 	at com.sun.corba.ee.impl.protocol.MessageMediatorImpl.doWork(MessageMediatorImpl.java:2193)
     [exec] 	at com.sun.corba.ee.impl.threadpool.ThreadPoolImpl$WorkerThread.performWork(ThreadPoolImpl.java:467)
     [exec] 	at com.sun.corba.ee.impl.threadpool.ThreadPoolImpl$WorkerThread.run(ThreadPoolImpl.java:510)
     [exec] Caused by: jakarta.ejb.CreateException: ERROR creating stateful SessionBean
     [exec] 	at com.sun.ejb.containers.StatefulSessionContainer.createRemoteBusinessObjectImpl(StatefulSessionContainer.java:662)
     [exec] 	at com.sun.ejb.containers.EJBHomeImpl.createRemoteBusinessObjectImpl(EJBHomeImpl.java:97)
     [exec] 	at com.sun.ejb.containers.EJBHomeInvocationHandler.invoke(EJBHomeInvocationHandler.java:158)
     [exec] 	... 18 more
     [exec] Caused by: java.lang.reflect.InvocationTargetException
     [exec] 	at com.sun.ejb.containers.BaseContainer.createEjbInstanceAndContext(BaseContainer.java:1710)
     [exec] 	at com.sun.ejb.containers.StatefulSessionContainer.createBeanInstance(StatefulSessionContainer.java:770)
     [exec] 	at com.sun.ejb.containers.StatefulSessionContainer.createRemoteBusinessObjectImpl(StatefulSessionContainer.java:649)
     [exec] 	... 20 more
     [exec] Caused by: java.lang.NullPointerException
     [exec] 	at java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:936)
     [exec] 	at org.jboss.weld.manager.BeanManagerImpl.getBean(BeanManagerImpl.java:1169)
     [exec] 	at org.jboss.weld.manager.BeanManagerImpl.getBean(BeanManagerImpl.java:201)
     [exec] 	at org.glassfish.weld.services.JCDIServiceImpl.createInterceptorInstance(JCDIServiceImpl.java:384)
     [exec] 	at com.sun.ejb.containers.BaseContainer.createEjbInterceptors(BaseContainer.java:1756)
     [exec] 	at com.sun.ejb.containers.BaseContainer.createEjbInstanceAndContext(BaseContainer.java:1687)
     [exec] 	... 22 more
```

manovotn commented:
> Thanks, although unfortunately Beta3 doesn't work either :(

Well, functionally, Weld 4 is equal to latest Weld 3, I am not sure why are you seeing this. But then again, I have no idea what previous version was GF run against but I suppose it must have worked with Weld 3?

> Still tons of exceptions when running the GF cdi_all test like this:

While I don't mind doing another release, it'd be nice if you executed those tests before asking for new release, e.g. run them against snapshot version of Weld. It'd be faster turnaround and we'd avoid several releases.

arjantijms commented:
>  But then again, I have no idea what previous version was GF run against but I suppose it must have worked with Weld 3?

The previous version that seems to work is 4.0.0.Alpha3. It started failing from 4.0.0.Beta1 onwards. To rule out my local environment, I'll do a PR with Beta3 for the GF repo, which will run the "cdi_all" test against our jenkins instance.

manovotn commented:
> The previous version that seems to work is 4.0.0.Alpha3.

There wasn't a functionality change between those versions that I can see causing this (in fact, there was almost no change at all). Although there was bunch of dependency updates some of which touched the OSGi bundling that you consume. So that's probably the issue? I'd start with comparison of the two bundles that it creates from the viewpoint of what GF needs to have inside and if there are any extras that you don't want.

lukasj commented:
@manovotn My bad, now I finally see where exactly I made a mistake when testing the fix for [WELD-2642](https://issues.redhat.com/browse/WELD-2642) which fixed one issue but introduced the issue we are seeing now - addition of missing export is correct, change in existing exports is wrong and that is the cause of what we see now, since package `org.jboss.weld.module` is not exported nor properly linked by other packages, so it is not visible to them.

When I was looking for the cause of the [WELD-2644](https://issues.redhat.com/browse/WELD-2644), I was going commit-by commit (... quite time-consuming thing...) and applying only addition of missing package to exports instead of applying full change as was done. That's why all tests on my end passed.

manovotn commented:
> since package org.jboss.weld.module is not exported nor properly linked by other packages

Ah, I think I can see what you mean now. The issue is that we actually have `org.jboss.weld.module` package inside `core/impl` but the OSGi reference was to our `core/modules` which has moved.

>  addition of missing export is correct, change in existing exports is wrong

So if I get you right, then you need something like [this](https://github.com/weld/core/pull/2159), correct?
I've set up a [Weld tracking issue](https://issues.redhat.com/browse/WELD-2645) as well.

Pandrex247 commented on annotations/2.0/apidocs/jakarta.annotation/jakarta/annotation/ManagedBean.html:
```diff
@@ -0,0 +1,268 @@
+<h4>value</h4>
+<pre><a href="https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html?is-external=true" title="class or interface in java.lang" class="externalLink">String</a>&nbsp;value</pre>
+<div class="block">The name of the Managed Bean. Managed Bean names must be unique within a
+ Jakarta EE module. For each named Managed Bean, Java EE containers must make
```
Is this a valid _Java EE_ reference?
Shouldn't it be `Jakarta EE containers`?

kwsutter replied:
@m0mus Let's clean this up at the same time as the other updates...  Thanks!

lukasj replied:
this is done too

lukasj commented:
@manovotn 

> > addition of missing export is correct, change in existing exports is wrong
> 
> So if I get you right, then you need something like [this](https://github.com/weld/core/pull/2159), correct?
> I've set up a [Weld tracking issue](https://issues.redhat.com/browse/WELD-2645) as well.

Yes, after this change, GF quicklook tests on my end are passing. Thanks!

manovotn commented:
Ok, I will get it released later today or first thing tomorrow and will let you know here once it's done.

manovotn commented:
Weld 4.0.0.Beta4 should now be in Central.

arjantijms commented:
@manovotn 

Maybe I'm crazy or missing something, but it still does not work with beta4. See also https://ci.eclipse.org/glassfish/job/glassfish_build-and-test-using-jenkinsfile/job/PR-23234/

At the moment I don't have much time to investigate. I guess this will at least bubble up when people are trying to release GlassFish 6, but hopefully I can find some time before then.

lukasj commented:
TBH, I'm running out of ideas on why it passed on my end :-/

lukasj commented:
@manovotn just noticed the difference in the content of `META-INF/services/org.jboss.weld.module.WeldModule`:

the 4-Alpha3 has:

```
org.jboss.weld.module.ejb.WeldEjbModule
org.jboss.weld.module.jta.WeldTransactionsModule
org.jboss.weld.module.web.WeldWebModule
```

while all betas do have only following entry there:

```
org.jboss.weld.module.web.WeldWebModule
```

When I copy `META-INF/services/org.jboss.weld.module.WeldModule` form Alpha3 to Beta4, GF tests are happy. This leads me to think there's some probably non-deterministic behaviour in the weld build

manovotn commented:
Hmm, that's awkward. When I build `weld-osgi-bundle-${VERSION}.jar` locally from the repo, I can see it contains all three entries, just like it should. Yet the release doesn't.
This should be performed by the [shade plugin](https://github.com/weld/core/blob/master/bundles/osgi/pom.xml#L184-L249), namely the transformer specified in there. But what the release had look like it instead of merging took just the last entry.
I'll try to reproduce that locally...

manovotn commented:
So, after a lot of tinkering, I am unable to reproduce this - the artifact always generates correctly.
What seems to be the issue is ordering of `bundle` and `shade` plugins which are both bound to `package` phase and according to their declaration order, they should execute with `bundle` going first and `shade` second. Which it does when you try it.

However, trying to execute a release, I can see that `release:perform` actually swaps these two steps which I suspect leads to the problem we are seeing. So far, I was unable to determine why this happens, the effective pom should be the same. I'll keep digging...

manovotn commented:
@lukasj @arjantijms There is now Beta5 which should have the `META-INF/services` specified correctly. See it in [Central](https://repo1.maven.org/maven2/org/jboss/weld/weld-osgi-bundle/4.0.0.Beta5/).
The tracking issue with details was [WELD-2646](https://issues.redhat.com/browse/WELD-2646). I haven't been able to nail down the cause and it almost looks like Maven issue, but the change I made should make it reliable again.

Give it a spin and let me know if we are finally good :-)

arjantijms commented:
@manovotn quite a relieve to learn I wasn't going crazy :) Good news, it now finally works! Both locally and on our CI:

https://ci.eclipse.org/glassfish/job/glassfish_build-and-test-using-jenkinsfile/job/PR-23234/

lukasj commented:
@manovotn Yes, this is it. It finally works for me on my end too. Thanks!

kwsutter commented:
* On ballot completion, the specification committee mentor:
 - [x] adds this final checklist to the main PR.
 - [x] adds the `approved` label to the PRs, and sends out the Ballot Summary per this [template](#ballot-summary-email-template) to the [public Jakarta EE Specification Committee email list](jakarta.ee-spec@eclipse.org)
 - [ ] calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job. Manually editing the jenkins Build Information will help identify the build (ie. Mail 2.0 or CDI 3.0).
 - [ ] merges the specification (and apidocs) PRs, ensuring the "date:" field in the _index.md file has an appropriate value to allow publishing.
 - [ ] updates the specification page with the ballot results. This is normally done via a separate PR that should be reviewed, approved, and merged.
 - [ ] notifies the EMO of the ballot results by email to [emo@eclipse-foundation.org](emo@eclipse-foundation.org). Just forward the ballot summary note sent earlier to the public Spec Committee email list.
 - [ ] creates an issue in the specification project that includes the following checklist for the specification project team:
 \<include url of Issue here>

<!-- Add this list to the specification project issue.  Delete this section of checklist from Spec PR once this specification project issue is created and logged. -->
 ----
 - [ ] promotes api staging release promotes the specification api jars to maven central. An example release job script can be found here https://wiki.eclipse.org/MavenReleaseScript.
 - [ ] go through the merged jakarta.ee specification website page to verify all the links are valid.
 - [ ] approve the compatibility request.
 - [ ] The compatible implementation project/vendor MUST send an email to tck@eclipse.org for approval of the compatible implementation for trademark usage.
 - [ ] merge any final release branch as appropriate for the branch management for the project.

kwsutter commented:
I need to remove the `approved` label for now while we figure out the next steps... I noticed that some of the artifacts were modified after the ballot started (staged API and javadoc).  These were done in reaction to @jeanouii's `-1` comments, so they are probably good.  But, we can't modify the artifacts in the middle of a vote.

lukasj commented:
@kwsutter I though this was on hold due to missing CI having final API "release" in it. To get such build, we had to fix issues in Weld first - that's why I hoped there is still window allowing addressing comments raised in this issue, sorry.

kwsutter commented:
> @kwsutter I though this was on hold due to missing CI having final API "release" in it. To get such build, we had to fix issues in Weld first - that's why I hoped there is still window allowing addressing comments raised in this issue, sorry.

No problem, @lukasj!  Just some miscommunications among several parties...  We'll get this straightened out.  I've been talking with @m0mus via Slack, so touch base with him and let me know when you are both ready to restart the ballot.  Thanks!

kwsutter commented:
I received a Slack message from @m0mus that this Spec PR is ready to go.  @lukasj, can I get confirmation from you as well?  From what I can tell, it looks ready (again).  But, I want to ensure that both of you are good with starting another ballot for Annotations. Thanks!

kwsutter reviewed (approved):
LGTM! Thanks, @m0mus and @lukasj!  Just waiting for a confirmation from Lukas before kicking off the ballot.

lukasj commented:
@kwsutter good to go from me as well. Thanks!

kwsutter commented:
* On ballot completion, the specification committee mentor:
 - [x] adds this final checklist to the main PR.
 - [ ] adds the `approved` label to the PRs, and sends out the Ballot Summary per this [template](#ballot-summary-email-template) to the [public Jakarta EE Specification Committee email list](jakarta.ee-spec@eclipse.org)
 - [ ] calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job. Manually editing the jenkins Build Information will help identify the build (ie. Mail 2.0 or CDI 3.0).
 - [ ] merges the specification (and apidocs) PRs, ensuring the "date:" field in the _index.md file has an appropriate value to allow publishing.
 - [ ] updates the specification page with the ballot results. This is normally done via a separate PR that should be reviewed, approved, and merged.
 - [ ] notifies the EMO of the ballot results by email to [emo@eclipse-foundation.org](emo@eclipse-foundation.org). Just forward the ballot summary note sent earlier to the public Spec Committee email list.
 - [ ] creates an issue in the specification project that includes the following checklist for the specification project team:
 \<include url of Issue here>

<!-- Add this list to the specification project issue.  Delete this section of checklist from Spec PR once this specification project issue is created and logged. -->
 ----
 - [ ] promotes api staging release promotes the specification api jars to maven central. An example release job script can be found here https://wiki.eclipse.org/MavenReleaseScript.
 - [ ] go through the merged jakarta.ee specification website page to verify all the links are valid.
 - [ ] approve the compatibility request.
 - [ ] The compatible implementation project/vendor MUST send an email to tck@eclipse.org for approval of the compatible implementation for trademark usage.
 - [ ] merge any final release branch as appropriate for the branch management for the project.
